clj-refactor.el

```diff
@@ -1983,37 +1983,68 @@ the alias in the project."
+  (assert (not (completable-for-cljr-slash? "a2.-")))
+  (assert (not (completable-for-cljr-slash? "a2.-a")))
+  (assert (not (completable-for-cljr-slash? "-")))
+  (assert (not (completable-for-cljr-slash? ".")))
+  (assert (not (completable-for-cljr-slash? "2")))
+  (assert (not (completable-for-cljr-slash? "+")))
+  (assert (not (completable-for-cljr-slash? "a/")))
+  (assert (not (completable-for-cljr-slash? "a/a")))
+  (assert (not (completable-for-cljr-slash? "a/a/"))))
+
 ;;;###autoload
 (defun cljr-slash ()
   "Inserts / as normal, but also checks for common namespace shorthands to require.
 If `cljr-magic-requires' is non-nil, executing this command after one of the aliases
 listed in `cljr-magic-require-namespaces', or any alias used elsewhere in the project,
 will add the corresponding require statement to the ns form."
   (interactive)
-  (insert "/")
-  (when-let (aliases (and cljr-magic-requires
-                          (not (cljr--in-map-destructuring?))
-                          (not (cljr--in-ns-above-point-p))
-                          (not (cljr--in-reader-literal-p))
-                          (not (cider-in-comment-p))
-                          (not (cider-in-string-p))
-                          (not (cljr--in-keyword-sans-alias-p))
-                          (not (cljr--in-number-p))
-                          (clojure-find-ns)
-                          (cljr--magic-requires-lookup-alias)))
-    (let ((short (cl-first aliases)))
-      (when-let (long (cljr--prompt-user-for "Require " (cl-second aliases)))
-        (when (and (not (cljr--in-namespace-declaration-p (concat ":as " short "\b")))
-                   (or (not (eq :prompt cljr-magic-requires))
-                       (not (> (length (cl-second aliases)) 1)) ; already prompted
-                       (yes-or-no-p (format "Add %s :as %s to requires?" long short))))
-          (save-excursion
-            (cljr--insert-in-ns ":require")
-            (let ((libspec (format "[%s :as %s]" long short)))
-              (insert libspec)
-              (ignore-errors (cljr--maybe-eval-ns-form))
-              (cljr--indent-defun)
-              (cljr--post-command-message "Required %s" libspec))))))))
+  (let ((original-input (cider-symbol-at-point)))
+    (insert "/")
+    (when-let (aliases (and cljr-magic-requires
+                            (completable-for-cljr-slash? original-input)
+                            (not (cljr--in-map-destructuring?))
+                            (not (cljr--in-ns-above-point-p))
+                            (not (cljr--in-reader-literal-p))
+                            (not (cider-in-comment-p))
+                            (not (cider-in-string-p))
+                            (not (cljr--in-keyword-sans-alias-p))
+                            (not (cljr--in-number-p))
+                            (clojure-find-ns)
+                            (cljr--magic-requires-lookup-alias)))
```

Comment:
> Btw, I always program on a desktop computer so this op is never slow for me, 

The story is probably more complex b/c I have a 16 core machine and cljr-slash takes 7s if not parallelized. So it can depend on "mystery factors" :)

Most of the slowness comes from refactor-nrepl. This PR is concerned with preventing refactor-nrepl from being queried at all - it's always  a nice improvement even when the backend is cached and parallelized.

> but one more thing we could do is to move the op from sync to async. 

This could plausibly create a race condition: one could press `/`, keep typing, get one's input inserted _and then_ be offered completions.

> One thing I often do that will get affected by this is type obscure-ns/a^H which will then pop up the completion candidates for that ns and I can then search for the thing I barely remember.

Mind to expand? Not sure I follow

Reply:
> Most of the slowness comes from refactor-nrepl. This PR is concerned with preventing refactor-nrepl from being queried at all - it's always a nice improvement even when the backend is cached and parallelized.

Definitely agree with that!  I was just thinking out loud about other ways we might bring the 7s down or avoid just sitting there waiting.

> This could plausibly create a race condition: one could press /, keep typing, get one's input inserted and then be offered completions.

Yeah that's a fair point, and I'm not sure what's worse: sitting around waiting for 7 seconds, or getting a prompt to narrow down a selection of candidates from a `/` you typed 7 seconds ago.  In my own projects I very rarely see the prompt because I'm really strict about consistent aliases.

> Mind to expand? Not sure I follow

Let's say I'm looking to make a `queue` and I remember that `medley` has a constructor, but I can't remember if it's `->queue` or simply `queue`.  So I type `medley/` and the lib is required and the ns form gets re-evaluated so all the symbols are available to the completion engine.  Now I just need to trigger the list of every symbol in that ns, so I type say `q` but I'm not sure the thing I'm looking for starts with `q` so I erase that letter and now I can hit `C-s` to search among the completion candidates.  If I search for `queue` I'd see both `queue` and `->queue`.
The problem is that none of this will work if the namespace hasn't yet been added to the ns form and re-evaluated.

Typing this all out made me realize that I've also been working around a weird quirk that `/` doesn't trigger completions on it's own, yet it's perfectly happy with displaying that same if I type one char and then remove it...
Does it work that way for you?

Reply:
> Most of the slowness comes from refactor-nrepl. This PR is concerned with preventing refactor-nrepl from being queried at all - it's always a nice improvement even when the backend is cached and parallelized.

Definitely agree with that!  I was just thinking out loud about other ways we might bring the 7s down or avoid just sitting there waiting.

> This could plausibly create a race condition: one could press /, keep typing, get one's input inserted and then be offered completions.

Yeah that's a fair point, and I'm not sure what's worse: sitting around waiting for 7 seconds, or getting a prompt to narrow down a selection of candidates from a `/` you typed 7 seconds ago.  Although, in my own projects I very rarely see the prompt because I'm really strict about consistent aliases.

> Mind to expand? Not sure I follow

Let's say I'm looking to make a `queue` and I remember that `medley` has a constructor, but I can't remember if it's `->queue` or simply `queue`.  So I type `medley/` and the lib is required and the ns form gets re-evaluated so all the symbols are available to the completion engine.  Now I just need to trigger the list of every symbol in that ns, so I type say `q` but I'm not sure the thing I'm looking for starts with `q` so I erase that letter and now I can hit `C-s` to search among the completion candidates.  If I search for `queue` I'd see both `queue` and `->queue`.
The problem is that none of this will work if the namespace hasn't yet been added to the ns form and re-evaluated.

Typing this all out made me realize that I've also been working around a weird quirk that `/` doesn't trigger completions on it's own, yet it's perfectly happy with displaying that same if I type one char and then remove it...
Does it work that way for you?

Reply:
> Yeah that's a fair point, and I'm not sure what's worse: sitting around waiting for 7 seconds, or getting a prompt to narrow down a selection of candidates from a / you typed 7 seconds ago.

The 7 seconds won't be a thing anymore following https://github.com/clojure-emacs/refactor-nrepl/pull/320 (and https://github.com/clojure-emacs/refactor-nrepl/issues/230#issuecomment-876645429 for cljs)

So both experiences (async vs. sync) would be as quick. But sync would avoid race conditions

>  Now I just need to trigger the list of every symbol in that ns, so I type say q but I'm not sure the thing I'm looking for starts with q so I erase that letter and now I can hit C-s to search among the completion candidates.

Personally I use the company-mode integration, so that I can see all names for a given prefix without having to type  a few chars tentatively:

![image](https://user-images.githubusercontent.com/1162994/128642059-85855589-23de-4f7f-9b4e-fe37d46b94c3.png)

I have to type `/ TAB`, so that's two different commands being composed. (I think that answer one of the questions)

Overall the cljr-slash integration works well for me - it successfully inserts a prefix and reloads the ns form